Migrate DQMEDAnalyzers from `stream` to `one` modules

fwyzard opened the pull request:
Migrate all DQMEDAnalyzer modules from `edm::stream::EDAnalyzer<...>` to `edm::one::EDAnalyzer<...>`.

This should significantly reduce the memory usage of a multithreaded DQM job, at the price of preventing the framework from analysing multiple runs or lumisections at the same time.
    
Lumisection-based histograms are handled in the `postModuleGlobalEndLumi` signal of the `DQMStore`.

If `DQMEDanalyzer` modules are migrated from `EDAnalyzer` to `EDProduce<Accumulator>`, that signal should be removed and the handling moved to the modules' `endLuminosityBlockProduce` method.

fwyzard committed:
Migrate DQMEDAnalyzers from stream to one modules

Migrate all DQMEDAnalyzer modules from `edm::stream::EDAnalyzer<...>` to
`edm::one::EDAnalyzer<...>`.

This should significantly reduce the memory usage of a multithreaded DQM
job, at the price of preventing the framework from analysing multiple
runs or lumisections at the same time.

Lumisection-based histograms are handled in the `endLuminosityBlockProduce`
method - which is only avaibale after the migration to `edm::one::EDProducer`.

For the time being, lumisection-based histograms are broken. If a
workaround is needed, one could use the `postModuleGlobalEndLumi` signal
of the `DQMStore` directly.

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
First attempt at a simpler, partial migration for #22157 .

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3364

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

DQMServices/Core


@vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please review it and eventually sign? Thanks.
@barvic this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

schneiml commented:
relative to #22157, I see
- pro: no mess with config changes.
- pro: no issues related to the new `edm::Accumulator` feature.
- con: these two were not the problems discussed in #22157.
- con: no idea if there is a disadavantage in using `EDAnalyzer`, what does @Dr15Jones think?

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3366

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #22218 was updated. @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3367

cmsbuild commented:
+1
Tested at: df166841f5249a8d1190cc709b42df100b20d434
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22218/26069/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22218/26069/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22218/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 32 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2465360
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2465182
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.1999999999 KiB( 22 files compared)
 - Checked 111 log files, 9 edm output root files, 27 DQM output files

schneiml commented:
This solution looks good. Since now fallback to stream modules will be needed, all the related code in `DQMEDAnalyzer`/`DQMStore` can be removed, as done here. There is still an issue with per-lumi histograms which can be fixed by adapting the output module.

I will assemble a new PR based on this commit, and split out the config changes of #22157 into #22243 . I will assemble another PR based on this commit and some more from #22157, plus the output module change, to migrate to `edm::one::EDAnalyzer`. These two are then orthogonal. Once both are in, it is to switch to `edm::one::EDProducer<Accumulator>`, if this is useful (@Dr15Jones ?). This can happen immediatly, or once we have the proper token-based dependencies implemented.

I suggest to close this PR.

fwyzard commented:
Mhm.

Changing the output modules will make it impossible to later support concurrent lumisections, also in global modules.

If I can think of a solution that preserves that possibility, I will push it here.

schneiml commented:
It is not like such changes can't be rolled back. At the moment, the goal is to keep the changes minimal and save for data taking, more changes will very likely follow over summer.

But yes, I you can come up with a solution, we are happy to use it!

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@schneiml , this should have a higher chance of working, if you want to have a look

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3414

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3414/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3414/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3420

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3420/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3420/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3421

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #22218 was updated. @cmsbuild, @civanch, @vazzolini, @kmaeshima, @mdhildreth, @dmitrijus, @nsmith-, @rekovic, @jfernan2, @kpedro88, @thomreis, @vanbesien can you please check and sign again.

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3425

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #22218 was updated. @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

cmsbuild commented:
+1
Tested at: a3159bbddd8f7a2fe7a2ca7437b97d98116b4e3d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22218/26132/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/0ef6a286b4b97552fb17e188750c4d2430276e67
https://github.com/cms-sw/cmssw/commit/cb25b50662fe3d1df89872e420c876d11497212b
https://github.com/cms-sw/cmssw/commit/4f0a2323f782044a40dfdcf8e973d35eb7c47fd3
https://github.com/cms-sw/cmssw/commit/239a1ba6aa93f823c13bb01b24193c8236418a0f
https://github.com/cms-sw/cmssw/commit/411fffdff01288f38b78e0e8b22ad7961c20f35a
https://github.com/cms-sw/cmssw/commit/2889abee905406acf037e1d2eef44acfc90aa90d
https://github.com/cms-sw/cmssw/commit/c7fa0d1be83fb5feca69eb9d761484e5cc68f36a
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22218/26132/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22218/26132/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22218/26132/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 135 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2465360
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2465190
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.71999999994 KiB( 22 files compared)
 - Checked 111 log files, 9 edm output root files, 27 DQM output files

schneiml commented:
@fwyzard do I see correctly that this PR is not safe to use atm (even though the tests pass), since there is no `edm::Accumulator` migration yet?

We will do the renaming now (since it should happen at some point anyways), and this PR can probably be closed, to avoid confusion.

fwyzard commented:
There is no migration from `Analyzer` to `Producer` here, hence possibility to use the `edm::Accumulator` feature.

fwyzard committed:
Handle lumisection-based histograms in the DQMStore

For the time being, lumisection-based histograms are cloned and set with
the correct lumisection index in the `postModuleGlobalEndLumi` signal
handler of the `DQMStore`.

If/when the `DQMEDAnalyzer` modules are migrated from `EDAnalyzer` to
`EDProducer<Accumulator>`, the call should be moved to the modules'
`endLuminosityBlockProduce` method.

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@schneiml @dmitrijus please have a look

What workflow can I use to validate the changes ?

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22218/3466

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #22218 was updated. @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

cmsbuild commented:
+1
Tested at: 5fe910bced3f015403e0684f415ff85898714c56
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22218/26190/summary.html

cmsbuild commented:
Comparison job queued.

dmitrijus commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22218/26190/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 32 differences found in the comparisons
 - DQMHistoTests: Total files compared: 28
 - DQMHistoTests: Total histograms compared: 2498032
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2497847
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.829999999951 KiB( 23 files compared)
 - Checked 115 log files, 9 edm output root files, 28 DQM output files

fabiocos commented:
+1

schneiml commented on DQMServices/Core/src/DQMStore.cc:
```diff
@@ -856,7 +707,7 @@ DQMStore::book_(const std::string &dir, const std::string &name,
     // Create and initialise core object.
     assert(dirs_.count(dir));
-    MonitorElement proto(&*dirs_.find(dir), name, run_, streamId_, moduleId_);
+    MonitorElement proto(&*dirs_.find(dir), name, run_, moduleId_);
```
@fwyzard ugh... What have you done there? Did you also remove the `streamId` from the `MonitorElement` constructor? It should be removed, but the change is not in this PR, so now `streamId`s and `moduleId`s are messed up!

fwyzard commented:
I thought I did fix all the constructors as well... I can have a look
tomorrow.

fwyzard commented:
@schneiml from a cursory look, I think I have updated all the calls to the `MonitorElement` constructors.

If you can point me where you see an inconsistency between the `streamId` and `moduleId` I will try to fix it.

schneiml commented:
@fwyzard streamId is still in the constructor declaration, so you are passing moduleId *instead* of streamId. Which would be sort of fine if it was used consistently, but some places do nontrival queries on the moduleId (e.g., `save()`). Just remove streamId from the constructor and see where it fails to compile...

fwyzard commented:
Ah, I see... and it was not spot before by the compiler because of all the default parameters.
OK, let's see if #22364 helps.
